sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala

```diff
@@ -305,12 +305,17 @@ object HiveTypeCoercion {
 
   /**
    * Convert all expressions in in() list to the left operator type
+   * except when the left operator type is NullType. In case when left hand
+   * operator type is NullType create a Literal(Null).
    */
   object InConversion extends Rule[LogicalPlan] {
     def apply(plan: LogicalPlan): LogicalPlan = plan resolveExpressions {
       // Skip nodes who's children have not been resolved yet.
       case e if !e.childrenResolved => e
 
+      case i @ In(a, b) if (a.dataType == NullType) =>
+        Literal.create(null, BooleanType)
```

Comment:
instead of just casting null to boolean, can we come up with a better idea according to the data types of `b`?